syntax = "proto3";

package graph;

service GraphService {
  rpc AddRelationship(AddRelationshipRequest) returns (AddRelationshipResponse);
  rpc GetRelationshipsForTerm(GetRelationshipsForTermRequest) returns (GetRelationshipsForTermResponse);
  rpc DeleteRelationship(DeleteRelationshipRequest) returns (DeleteRelationshipResponse);
}

enum RelationshipType {
  UNKNOWN = 0;
  RELATED_TO = 1;
  IS_A = 2;
  CONTAINS = 3;
  DEPENDS_ON = 4;
}

message Relationship {
  string from_term_id = 1;
  string to_term_id = 2;
  RelationshipType type = 3;
}

message AddRelationshipRequest {
  string from_term_id = 1;
  string to_term_id = 2;
  RelationshipType type = 3;
}

message AddRelationshipResponse {
  bool success = 1;
}

message GetRelationshipsForTermRequest {
  string term_id = 1;
}

message GetRelationshipsForTermResponse {
  repeated Relationship relationships = 1;
}

message DeleteRelationshipRequest {
  string from_term_id = 1;
  string to_term_id = 2;
  RelationshipType type = 3;
}

message DeleteRelationshipResponse {
  bool success = 1;
}